caas/specs/rbac.go

```diff
@@ -78,36 +81,38 @@ func (sa ServiceAccountSpecV3) Validate() error {
 			return errors.Trace(err)
 		}
 	}
-	return nil
+	roleNames := set.NewStrings()
+	for _, r := range sa.Roles {
+		if err := r.Validate(); err != nil {
+			return errors.Trace(err)
+		}
+		if r.Name == "" {
+			continue
+		}
+		if roleNames.Contains(r.Name) {
+			return errors.NotValidf("duplicated role name %q", r.Name)
+		}
+		roleNames.Add(r.Name)
+	}
+	if roleNames.Size() == 0 || len(sa.Roles) == roleNames.Size() {
+		// All good.
+		return nil
+	}
+	return errors.NewNotValid(nil, "either all or none of the roles should have a name set")
 }
 
 // PrimeServiceAccountSpecV3 defines spec for creating the prime RBAC resources for version 3.
 type PrimeServiceAccountSpecV3 struct {
-	name                 string
+	Name                 string
```

Comment:
we now allow charm to set primary `SA` name?

Reply:
ah good point, I think I should go back to making it unexported but not have the SetName() also set the role name